packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+
+    it("can use getDeviceSources", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const emptyArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+
+        expect(emptyArray.length).toBe(0);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const touchArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+        expect(touchArray.length).toBe(2);
+    });
+
+    it("can use onDeviceConnectedObservable", () => {
+        expect.assertions(1);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        let observableSource = null;
+
+        deviceSourceManager.onDeviceConnectedObservable.add((eventData) => {
+            observableSource = eventData;
+        });
+
+        deviceSourceManager._addDevice(touchSource);
+
+        expect(observableSource).toEqual(touchSource);
+    });
+
+    it("can use onDeviceDisconnectedObservable", () => {
+        expect.assertions(2);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        let observableSource = null;
+
+        deviceSourceManager.onDeviceDisconnectedObservable.add((eventData) => {
+            observableSource = eventData;
+        });
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._removeDevice(DeviceType.Touch, 0);
+
+        const nullSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 0);
+
+        expect(observableSource).toEqual(touchSource);
+        expect(nullSource).toBe(null);
+    });
+
+    it("can talk to DeviceSource onInputChangedObservable", () => {
+        expect.assertions(2);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        let observableEvent = null;
+
+        mouseSource.onInputChangedObservable.add((eventData) => {
+            observableEvent = eventData;
+        });
+
+        const mouseEvent: IUIEvent = { pointerId: 0, type: "pointerdown", pointerType: "mouse", clientX: 1, clientY: 1 } as IPointerEvent;
+
+        // Try to call _onInputChanged without any added DeviceSources
+        deviceSourceManager._onInputChanged(DeviceType.Mouse, 0, mouseEvent);

```

Comment:
This is a strange why to test this. We should hooking in only at the deviceInputSystem level and not anywhere else. Let's chat a bit offline.

Reply:
I'll change this as per our offline conversation